Token State Issue Fix and Misc Bugfixes

dternyak opened the pull request:
## Bugs

1.   User could input duplicate tokens via custom token form. This could cause a corruption to state that results in potentially incorrect tokens being sent via subsequent transactions.
2.   User could cause unexpected state to occur by deleting a custom token that has a corresponding default token. This could cause a corruption to state that results in potentially incorrect tokens being sent via subsequent transactions. 


## Methods of Encounter
1. User adds a custom token that is already in local storage.
2. User adds a a custom token that is already in the default token list.
3. MEW updates the default token list with a token(s) that conflicts with a user’s existing custom token(s).

## Solutions
1. Prevent user from adding a custom token with the same symbol and network as an existing default or custom token.
2. Remove custom tokens that conflict with default tokens (matching symbol or contract address).
3. De-dup custom tokens that match with another custom token (only possible before this PR, as (1) prevents this from happening in the future).
4. Prevent user from adding a custom token with the same contract address and network as an existing default or custom token.

##### Additionally, a handful of additional fixes were implemented alongside the token state corruption fix:
1. Network type is now set on new custom tokens.
2. Node selection change will cause a refresh when networks change (to ensure clean state).
3. Deleting a custom token will hide rawTx/signedTx textAreas (rootScope hack, but doesn’t introduce significant complexity).
4. Reset contractAddress input after successfully saving a token.

Commit:
Fix token conflict bugs

dternyak committed:
additional error handling; remove console logs; de-duplicate any existing custom tokens

dternyak committed:
formatting and dead code removal

dternyak committed:
formatting; rename vars; remove unused function

dternyak committed:
check if new custom token already exists in defaultTokens before persisting to state in addition to de-duping custom tokens that conflict with default tokens

skubakdj committed:
clean original PR diff by reverting walletBalanceCtrl.js changes

wbobeirne commented on app/scripts/myetherwallet.js:
```diff
@@ -21,19 +21,111 @@ Wallet.generate = function(icapDirect) {
 }
+
+function deDup(list) {
+  var uniq = new Set(list.map(e => JSON.stringify(e)));
```
Not sure what our support game is, but Sets are pretty new: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#Browser_compatibility. Fortunately there are many ways to skin the cat of deduping an array.

dternyak replied:
`removeDuplicates ` will handle this anyway, so I'll just remove.

wbobeirne reviewed (commented):
My general sense is it's a little difficult to tell when something is being disqualified for conflicting symbol vs conflicting contract address. But this looks pretty thorough for fixing our bug. I haven't tested myself though, so I don't want to drop in an approve.

dternyak committed:
Merge pull request #1 from skubakdj/clean-PR-diff

clean up PR diff

dternyak committed:
completely undo whitespace changes in app/scripts/controllers/walletBalanceCtrl.js

dternyak committed:
merge and fix conflicts

dternyak committed:
remove unneeded generic de-dup

dternyak committed:
reset contract address input after succesfully saving custom token

dternyak committed:
only reload on node selection change when networks change

dternyak committed:
Hide rawTx/signedTx text areas after custom token deletion via rootScope.

dternyak committed:
Standardize on removing duplicates via symbol.

skubakdj committed:
prevent duplicate token contract addresses

skubakdj committed:
incorporate custom token name into error message

dternyak committed:
Merge pull request #2 from skubakdj/Token-conflict-Fix

Prevent duplicate contract addresses

dternyak committed:
- de-dup existing custom tokens via contractAddress
- simply removeConflictingTokensFromLocalStorage
- rename various function to better describe return type (bool)
